terra_notebook_utils/table.py

```diff
@@ -1,10 +1,131 @@
+# assert types.pop() in VALUE.__args__, f"Row: {row.name}: List members must be of type {VALUE.__args__}"
+            else:
+                pass
+                # raise TypeError(f"Row {row.name}: data table value must be of type{VALUE.__args__}")
+            request_data.extend(update_ops)
+        return request_data
+
+    def put_row(self, item: Union[ROW_LIKE, ATTRIBUTES]) -> str:
+        if isinstance(item, dict):
+            row = Row(f"{uuid4()}", item)
+        else:
+            row = Row(*item)
+        column_headers = tuple(sorted(row.attributes.keys()))
+        if column_headers not in self._tsvs:
+            self._tsvs[column_headers] = "\t".join([f"{self.name}_id", *column_headers])
+            self._row_update_request_data[column_headers] = list()
+        self._tsvs[column_headers] += (
+            os.linesep
+            + row.name + "\t"
+            + "\t".join(row.attributes[c] if isinstance(row.attributes[c], str)
+                        else "x"
+                        for c in column_headers)
+        )
+        update_request_data = self._get_row_update_request_data(row)
+        if update_request_data:
+            self._row_update_request_data[column_headers].append((row, update_request_data))
+        self._upload(1024 * 20)
+        return row.name
+
+    def _upload(self, threshold: Optional[int]=None):
+        """
+        Schedule uploads for all TSVs of size equal to or greater than `threshold`.
+        If `threshold` is None, schedule uploads for all TSVs.
+        """
+        for column_headers, tsv in self._tsvs.copy().items():
+            if threshold is None or len(tsv) >= threshold:
+                row_updates = self._row_update_request_data.get(column_headers, list())
+                self.submit(self._do_fiss_upload, tsv, row_updates)
+                del self._tsvs[column_headers]
+                if row_updates:
+                    del self._row_update_request_data[column_headers]
+
+    def _do_fiss_upload(self, tsv: str, row_update_request_data: List[Tuple[Row, List[Dict[str, Any]]]]):
+        fiss.fapi.upload_entities(self._workspace_google_project,
+                                  self._workspace_name,
+                                  tsv,
+                                  model="flexible").raise_for_status()
+        for row, request_data in row_update_request_data:
+            if request_data:
+                self.submit(self._do_fiss_updates, row, request_data)
+
+    def _do_fiss_updates(self, row: Row, request_data: UPDATE_OPS, retry: int=0):
+        try:
+            fiss.fapi.update_entity(self._workspace_google_project,
+                                    self._workspace_name,
+                                    self.name,
+                                    row.name,
+                                    request_data).raise_for_status()
+        except requests.exceptions.HTTPError as e:
+            if 500 == e.response.status_code:
```

Comment:
Would `if e.response.status_code >= 500:` or `if e.response.status_code in (500, 502, 503, 504):` make sense here?

Reply:
Possibly. But without knowledge of what `50x` errors means, it’s difficult to know what to do.

For instance, this line handles the possibility of a successful request returning `500`. This is pathological behavior from Firecould. What should we do with `50x` errors? Who knows.

So you may be right. However, I’d prefer to handle the error narrowly until there’s evidence of pathological `50x` errors.